src/graphics/program-lib/programs/standard.js

```diff
@@ -1391,6 +1445,120 @@ var standard = {
+                    DecodeClusterFloat2(colorB.xy)
+                ) * clusterCompressionLimit0.y;
+    
+                // spot light cos angles
+                vec4 coneAngle = texture2D(lightsTexture8, vec2(3.5 * lightsTextureInvSize.z, lightV));
+                float innerConeAngleCos = DecodeClusterFloat2(coneAngle.xy) * 2.0 - 1.0;
+                float outerConeAngleCos = DecodeClusterFloat2(coneAngle.zw) * 2.0 - 1.0;
+    
+    
+                #ifdef CLUSTER_TEXTURE_FLOAT
+    
+                    vec4 lightPosRange = texture2D(lightsTextureFloat, vec2(0.5 * lightsTextureInvSize.x, lightV));
+                    vec3 lightPos = lightPosRange.xyz;
+                    float range = lightPosRange.w;
+    
+                    // spot light direction
+                    vec4 lightDir_Unused = texture2D(lightsTextureFloat, vec2(1.5 * lightsTextureInvSize.x, lightV));
+                    vec3 lightDir = lightDir_Unused.xyz;
+    
+                #else   // 8bit
+    
+                    vec4 encPosX = texture2D(lightsTexture8, vec2(4.5 * lightsTextureInvSize.z, lightV));
+                    vec4 encPosY = texture2D(lightsTexture8, vec2(5.5 * lightsTextureInvSize.z, lightV));
+                    vec4 encPosZ = texture2D(lightsTexture8, vec2(6.5 * lightsTextureInvSize.z, lightV));
+    
+                    vec3 lightPos = vec3(
+                        DecodeClusterFloat4(encPosX),
+                        DecodeClusterFloat4(encPosY),
+                        DecodeClusterFloat4(encPosZ)
+                    ) * clusterBoundsDelta + clusterBoundsMin;
+    
+                    vec4 encRange = texture2D(lightsTexture8, vec2(7.5 * lightsTextureInvSize.z, lightV));
+                    float range = DecodeClusterFloat4(encRange) * clusterCompressionLimit0.x;
+    
+                    // spot light direction
+                    vec4 encDirX = texture2D(lightsTexture8, vec2(8.5 * lightsTextureInvSize.z, lightV));
+                    vec4 encDirY = texture2D(lightsTexture8, vec2(9.5 * lightsTextureInvSize.z, lightV));
+                    vec4 encDirZ = texture2D(lightsTexture8, vec2(10.5 * lightsTextureInvSize.z, lightV));
+    
+                    vec3 lightDir = vec3(
+                        DecodeClusterFloat4(encDirX),
+                        DecodeClusterFloat4(encDirY),
+                        DecodeClusterFloat4(encDirZ)
+                    ) * 2.0 - 1.0;
+    
+                #endif
+    
+                // evaluate point light
+                getLightDirPoint(lightPos);
+                dAtten = getFalloffLinear(range);
+                if (dAtten > 0.00001) {
+    
+                    // spot light
+                    if (lightType > 0.5) {
+                        dAtten *= getSpotEffect(lightDir, innerConeAngleCos, outerConeAngleCos);
+                    }
+    
+                    dAtten *= getLightDiffuse();
+                    dDiffuseLight += dAtten * lightColor;
+                }
+            }
```

Comment:
Could this be a chunk?

Reply:
Apart from `${clusterTextureFormat}` it does seem suitable for putting into a chunk - this differs significantly from the old lighting loop which was generated mostly in javascript.

Reply:
done this part. I guess I was hesitant to do it as it's not finished (all light types, specular, other light properties) .. and it felt like committing it to a chunk would make it harder to change as people could have overwritten it. But that is pretty unlikely.